iceoryx_hoofs/include/iceoryx_hoofs/internal/cli/command_line_argument_parser.hpp

```diff
@@ -0,0 +1,103 @@
+///        for consistency and syntax checks.
+class CommandLineArgumentParser
+{
+  public:
+    static constexpr uint64_t OPTION_OUTPUT_WIDTH = 45;
+
+  private:
+    friend class OptionManager;
+    friend CommandLineOptionValue
+    parseCommandLineArguments(const CommandLineOptionSet&, int, char*[], const uint64_t, const UnknownOption) noexcept;
+
+    /// @brief Parses the arguments from the command line.
+    ///        Calls onFailureCallback in optionSet when the command line arguments contain illegal syntax or required
+    ///        values are not provided and prints the help.
+    /// @param[in] optionSet the user defined options, based on those options the CommandLineOptionValue object is
+    ///            generated
+    /// @param[in] argc number of arguments, see int main(int argc, char*argv[])
+    /// @param[in] argv the string array of arguments, see int main(int argc, char*argv[])
+    /// @param[in] argcOffset the starting point for the parsing. 1U starts at the first argument.
+    /// @param[in] actionWhenOptionUnknown defines the action which should be performed when the user sets a
+    ///             option/switch which is unknown
+    CommandLineOptionValue parse(const CommandLineOptionSet& optionSet,
+                                 int argc,
+                                 char* argv[],
+                                 const uint64_t argcOffset = 1U,
+                                 const UnknownOption actionWhenOptionUnknown = UnknownOption::TERMINATE) noexcept;
+
+    void printHelpAndExit() const noexcept;
+
+    /// BEGIN only used in parse to improve readability
+    ///
+    /// Do not use those internal methods outside of parse. They were written
+    /// to improve the readability of the code. None of those functions verify
+    /// the pre conditions they require, this has to be done by calling them
+    /// in the correct order.
+    bool areAllRequiredValuesPresent() const noexcept;
+    bool hasArguments(const int argc) const noexcept;
+    bool assignBinaryName(const char* name) noexcept;
+    bool doesOptionStartWithMinus(const char* option) const noexcept;
+    bool hasOptionName(const char* option) const noexcept;
+    bool hasValidSwitchName(const char* option) const noexcept;
+    bool hasValidOptionName(const char* option) const noexcept;
+    bool doesOptionNameFitIntoString(const char* option) const noexcept;
+    bool isNextArgumentAValue(const uint64_t position) const noexcept;
+    bool isOptionSet(const CommandLineOptionSet::Value& entry) const noexcept;
+    bool doesOptionValueFitIntoString(const char* value) const noexcept;
+    bool doesOptionHasSucceedingValue(const CommandLineOptionSet::Value& entry, const uint64_t position) const noexcept;
+    /// END only used in parse to improve readability
+
+    void setDefaultValuesToUnsetOptions() noexcept;
+
+  private:
+    int m_argc = 0;
+    char** m_argv = nullptr;
+    uint64_t m_argcOffset = 0;
```

Comment:
I wouldn't make it const since I set all those variable inside of `parse` and require a late initialization - because of the macro construct.
Another alternative would be to pack it into a optional but then I have to check if its set.

So `uint64_t` for `m_argc` but no const.